### PR TITLE
TestNG 7.4.0 Can not skip test through listener

### LINKED ISSUE #2522: TestNG 7.4.0  Can not skip test through listener
It looks like  testng  continue execute test   even there is skip exception has been generated  through listener

```java
public class TestListener implements  IInvokedMethodListener {
  @Override
    public void beforeInvocation(IInvokedMethod invokedMethod, ITestResult result) {
            if (method.getName().contains("oneTest"))
            throw new SkipException("skip test ...");
   }
   
  @Override
    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
    }
}
```

```java
@Listeners({TestListener.class})
public class TestSample  {

    @Test()
    public void oneTest() {
        System.out.println("aaaaaaa");
        Assert.assertEquals(1,2);
    }

    @Test()
    public void twoTest() {
        Assert.assertEquals(1,1);
    }
}
```

when run TestSample ,  
```  
[Utils] [WARN] [Warning] Caught a [SkipException] exception from one of listeners com.sample.tests.TestListener. Will mark [com.sample.tests.TestSample.oneTest()] as SKIPPED.
aaaaaaa
.....
....
......
```

testNG still continue execute code .

`BaseInvoker.java` catch the exception , but does not handle it properly ...  

```java
  InvokedMethodListenerInvoker invoker =
        new InvokedMethodListenerInvoker(listenerMethod, testResult, testResult.getTestContext());
    for (IInvokedMethodListener currentListener : m_invokedMethodListeners) {
      try {
        invoker.invokeListener(currentListener, invokedMethod);
      } catch (SkipException e) {
        String msg = String.format(
            "Caught a [%s] exception from one of listeners %s. Will mark [%s()] as SKIPPED.",
            SkipException.class.getSimpleName(), currentListener.getClass().getName(),
            invokedMethod.getTestMethod().getQualifiedName());
        Utils.warn(msg);
        testResult.setStatus(ITestResult.SKIP);
        testResult.setThrowable(e);
      }
    }
```

### PR DESCRIPTION
**Pull request by bj-9527**
TestNG 7.4.0 Can not skip test through listener
Fixes #2522 .

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@bj-9527 

Couple of things.

* I remember you had earlier raised another PR on the similar lines as part of https://github.com/cbeust/testng/pull/2423 Is this PR addressing the same problem (A test is not being skipped via a `SkipException` being thrown from a listener) or is this addressing something else ?
* Can we resurrect the original PR and then work on getting that closed (or) you would like to do a fresh start using this PR?
* There are merge conflicts which needs to be fixed via a fetch/rebase

**Comment by bj-9527** [comment]
Hi @krmahadevan 

I checked the previous PR  #2423, and compare them with this one, then I think this solution would be better. So I prefer use this one to solve the issue.
I have merged both test cases in each PR, and they are all works fine, and I also rebase the code and please help to review.

**Review by juherr — CHANGES_REQUESTED** [review]
Some cosmetic remarks on tests

File: core/src/test/java/test/listeners/github2522/TestListener.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+        status.put(result.getTestClass().getName() + result.getName(), result.getStatus());
+        if (hasFailures) {
+            throw new SkipException("SkipException");
+        }
+    }
+
+    @Override
+    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
+        status.put(testResult.getTestClass().getName() + testResult.getName(), testResult.getStatus());
```
[the remark this answers is not in the document]
Comment (bj-9527) [reply core/src/test/java/test/listeners/github2522/TestListener.java]: > Some cosmetic remarks on tests

@juherr Thanks a lot for your comments, I did several change based on that, please help to check.

**Comment by bj-9527** [comment]
@juherr Thanks a lot for your comments, I did several change based on that, please help to check.

File: core/src/test/java/test/listeners/github2522/IssueTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package test.listeners.github2522;
+
+import org.testng.Assert;
+import org.testng.ITestResult;
+import org.testng.TestNG;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+import test.SimpleBaseTest;
+
+@Listeners({TestListener.class})
```
Comment (juherr) [inline core/src/test/java/test/listeners/github2522/IssueTest.java:10]: You can remove the listener

Comment (bj-9527) [reply core/src/test/java/test/listeners/github2522/IssueTest.java]: OK, this have been done

File: core/src/test/java/test/listeners/github2522/SkipTestListener.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import org.testng.IInvokedMethod;
+import org.testng.IInvokedMethodListener;
+import org.testng.ITestResult;
+import org.testng.SkipException;
+
+public class SkipTestListener implements IInvokedMethodListener {
+    @Override
+    public void beforeInvocation(IInvokedMethod invokedMethod, ITestResult result) {
+        if (invokedMethod.getTestMethod().getMethodName().contains("oneTest")) {
+            throw new SkipException("skip test ...");
```
Comment (juherr) [inline core/src/test/java/test/listeners/github2522/SkipTestListener.java:12]: It is not recommended to throw exceptions in listeners.
Could you replace it with `result.setStatus(SKIP)` or add another test case.

**Review by juherr — APPROVED** [review]
LGTM but exceptions are not recommended in listeners.

**Review by krmahadevan — APPROVED** [review]

**Pull request merged**